src/utils/retry.ts

```diff
@@ -0,0 +1,271 @@
+  context?: RetryLogContext
+  shouldAbort?: (error: unknown) => boolean
+}
+
+const defaultOptions: Partial<PRetryOptions> = {
+  retries:
+    process.env.P_RETRY_RETRIES != null
+      ? (() => {
+          const val = Number(process.env.P_RETRY_RETRIES)
+          return Number.isFinite(val) && val >= 0 ? val : 5
+        })()
+      : isTest
+        ? 1
+        : 5,
+  factor: 2,
+  minTimeout: isTest ? 0 : 500,
+  maxTimeout: isTest ? 0 : 5000,
+  randomize: !isTest
+}
+
+function coerceStatus(val: any): number | undefined {
+  if (val === undefined || val === null) return undefined
+  const n = Number(val)
+  return Number.isFinite(n) ? n : undefined
+}
+
+export function getStatus(err: any): number | undefined {
+  if (!err || typeof err !== 'object') return undefined
+
+  // Handle shapes from Octokit/axios/fetch-like wrappers and coerce to number
+  const raw =
+    err.status ??
+    err.response?.status ??
+    err.response?.statusCode ??
+    err.statusCode ??
+    err.cause?.status ??
+    err.cause?.response?.status
+  return coerceStatus(raw)
+}
+
+function getHeaders(err: any): Record<string, string> | undefined {
+  // Normalize to a plain object of lower-cased header names
+  let headers = err?.headers ?? err?.response?.headers
+
+  // Handle Headers object (like from fetch)
+  if (!headers && err?.response?.headers?.get && err?.response?.headers?.keys) {
+    try {
+      headers = Object.fromEntries(
+        Array.from(err.response.headers.keys()).map((k: string) => [
+          k,
+          err.response.headers.get(k) as string
+        ])
+      )
+    } catch {
+      headers = undefined
+    }
+  }
+
+  if (!headers) return undefined
+  const out: Record<string, string> = {}
+  for (const [k, v] of Object.entries(headers)) {
+    out[k.toLowerCase()] = String(v)
+  }
+  return out
+}
+
+function errorToMessage(err: any): string {
+  const msg = err?.message ?? String(err)
+  const status = getStatus(err)
+  return status ? `[${status}] ${msg}` : msg
+}
+
+function shouldAbortDefault(err: any): boolean {
+  const status = getStatus(err)
+  const headers = getHeaders(err) || {}
+
+  // Retry-after and rate limit handling
+  if (status === 429) return false // always retry Too Many Requests
+
+  if (status === 403) {
+    const retryAfter = headers['retry-after']
+    const remaining = headers['x-ratelimit-remaining']
+    if (retryAfter || remaining === '0') return false // rate limited -> retry
+    return true // other 403s likely auth/permission -> abort
+  }
+
+  if (status == null) return false // network/unknown -> retry
+  if (status >= 500) return false // 5xx -> retry
+  if (status >= 400 && status < 500) return true // other 4xx -> abort
+  return false
+}
+
+export async function withRetry<T>(
+  fn: () => Promise<T>,
+  options: WithRetryOptions = {}
+): Promise<T> {
+  const { context, shouldAbort, ...retryOpts } = options
+
+  return pRetry(
```

Comment:
Il me semble que les `(X as any)` dans cette fonction ne sont pas utiles.
Ça rendra les choses un poil plus claires déjà

Reply:
En effet, j'ai enlevé tous ceux que je pouvais (un seul était nécessaire).

J'ai déjà fait ce que je pouvais pour alléger cette fonction au cours des itérations, mais pour qu'elle reste fonctionnelle je n'ai pas de meilleur idée pour le moment, après si t'es inspiré, you're welcome ;-)